### ISSUE TITLE
Does this support torch.nn.parallel.DistributedDataParallel?

### ISSUE THREAD

**Issue by acgtyrant** [vacancy/Synchronized-BatchNorm-PyTorch#1]

**Comment by vacancy** [comment]
Currently not.

The implementation is designed for multi-GPU BatchNorm, which is commonly used for Computer Vision tasks. Thus, it uses NCCL for multi-GPU broadcasting and reduction.

Distributed version needs other types of synchronization primitive operations (e.g., via shared memory or cross-machine synchronization).

Contributions will be highly appreciated!

**Comment by zhanghang1989** [comment]
Just let you know that PyTorch Compatible Synchronized Batch Norm is provided here http://hangzh.com/PyTorch-Encoding/index.html 
See the example [here](https://github.com/zhanghang1989/PyTorch-SyncBatchNorm).

**Comment by acgtyrant** [comment]
@zhanghang1989 Does this support torch.nn.parallel.DistributedDataParallel?

**Comment by zhanghang1989** [comment]
@acgtyrant yes, see the mnist example https://github.com/zhanghang1989/PyTorch-SyncBatchNorm#mnist-example

**Comment by acgtyrant** [comment]
@zhanghang1989 Excuse me, I only see that you use the `DataParallel`, not the `DistributedDataParallel`. If you are sure, I use the `DistributedDataParallel` by myself later.

**Comment by acgtyrant** [comment]
BTW, the python notebook is 404.

**Comment by zhanghang1989** [comment]
This no reason that it won't work with `DistributedDataParallel `.  https://github.com/zhanghang1989/PyTorch-SyncBatchNorm/blob/master/mnist.py
I will add  python notebook later

**Comment by vacancy** [comment]
@zhanghang1989 Hi Hang. Thanks for the introduction. This repo aims at providing a standalone and easy-to-use version of sync_bn such that it can be easily integrated into any existing frameworks. Its implementation also differs from your previous implementation.

For an example use of the sync_bn, please check:
https://github.com/CSAILVision/semantic-segmentation-pytorch

As for the DistributedDataParallel, currently, I have no plan for supporting it. @zhanghang1989 Have you tested your implementation in the distributed setting?

**Comment by zhanghang1989** [comment]
Thanks @vacancy for the introduction! Nice work. I have plan of supporting distributed training, due to a recent paper use it in object detection. 
@acgtyrant The only thing need to be considered for distribute is making sure the number of gpus is set correctly https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/nn/syncbn.py#L196

**Comment by vacancy** [comment]
@zhanghang1989 I am not an expert at this. But it seems to me that the DistributedDataParallel uses different implementation as DataParallel for broadcasting and reduction. While in a single process, multi-thread setting (DataParallel), one can use simple NCCL's `broadcast` and `reduce`, for tensors shared across multiple *processes* or even multiple *machines*, we need special implementation, which is defined in the package `torch.distributed`.

**Comment by zhanghang1989** [comment]
@vacancy Thanks for the information.

**Comment by acgtyrant** [comment]
I read the source code of `DistributedDataParallel`, I find that it does not broadcast the parameters like `DataParallel`, it only all_reduce the gradient so that all model_replicates use the same gradient to optimize the same model at all, and all model_replicates use the same buffer which is broadcasted from the model in the device 0 of the rank 0 node.

@zhanghang1989 I think your syncbn use only `comm.broadcast_coalesced` and `comm.reduce_add_coalesced` which only support cross-gpu, it does not support cross-node which is distributed.

**Comment by acgtyrant** [comment]
I wrote a distributed bn, but it does not work, any idea?

```
import torch
from torch.autograd import Function
import torch.cuda
import torch.nn.functional as F
from torch.nn.modules.batchnorm import _BatchNorm


class AllReduce(Function):
    @staticmethod
    def forward(ctx, input_):
        torch.distributed.deprecated.all_reduce(input_)
        return input_

    @staticmethod
    def backward(ctx, grad_output):
        torch.distributed.deprecated.all_reduce(grad_output)
        return grad_output


class SynchronizedBatchNorm2d(_BatchNorm):
    def __init__(
            self, num_features, eps=1e-5, momentum=0.1, affine=True,
            track_running_stats=True):
        super(SynchronizedBatchNorm2d, self).__init__(
                num_features, eps, momentum, affine,
                track_running_stats)
        self.world_size = torch.distributed.deprecated.get_world_size()

    def forward(self, input_):
        if input_.dim() != 4:
            raise ValueError('expected 4D input (got {}D input)'
                             .format(input_.dim()))

        if not self.training:
            return F.batch_norm(
                input_, self.running_mean, self.running_var,
                self.weight, self.bias,
                self.training, self.momentum, self.eps)

        batchsize, channels, height, width = input_.size()
        numel = batchsize * height * width
        input_ = input_.permute(1, 0, 2, 3).contiguous().view(channels, numel)
        sum_ = input_.sum(1)
        sum_of_square = input_.pow(2).sum(1)
        sum_ = AllReduce.apply(sum_)
        sum_of_square = AllReduce.apply(sum_of_square)
        mean = sum_ / (numel * self.world_size)
        sumvar = sum_of_square - sum_ * mean
        unbias_var = sumvar / (numel * self.world_size - 1)
        bias_var = sumvar / (numel * self.world_size)
        std = bias_var.clamp(self.eps) ** 0.5

        self.running_mean = (
                (1 - self.momentum) * self.running_mean
                + self.momentum * mean.detach())
        self.running_var = (
                (1 - self.momentum) * self.running_var
                + self.momentum * unbias_var.detach())

        if self.affine:
            output = (
                    (input_ - mean.unsqueeze(1)) / std.unsqueeze(1) *
                    self.weight.unsqueeze(1) + self.bias.unsqueeze(1))
        else:
            output = (input_ - mean.unsqueeze(1)) / std.unsqueeze(1)

        return output.permute(1, 0).contiguous().view(
                batchsize, channels, height, width)
```

Thank you.

**Comment by vacancy** [comment]
Hi @acgtyrant 

What do you exactly mean by "SynchronizedBatchNorm2d is not numerical stable only"?

**Comment by acgtyrant** [comment]
My tests show that my implemented SynchronizedBatchNorm2d is not numerical stable as your bn, there is not any other error. So now that your bn can work, my bn should work too. But it does not, I do not know why... So I need some help now.

**Comment by acgtyrant** [comment]
After fix the wrong view of the output, I retrain drn now, and it seems as expected.

![2018-11-02-152419_1914x956_scrot](https://user-images.githubusercontent.com/3921062/47899106-6a602c00-deb3-11e8-933c-6ff876e09616.png)

**Comment by Cadene** [comment]
@acgtyrant Is DistributedDataParallel working for you? Are you planning to send a Pull Request?

Thank you all for this beautiful work!

**Comment by acgtyrant** [comment]
My distributed synced bn works in DistributedDataParallel, but because of the secrecy laws from my boss, I delete the post which contains the source of the distributed synced bn, sorry.

However, it is easy to be implemented, the torch.distributed.all_reduce is synced automatically, so just warp it as an autograd function and use it to all reduce sum and sum_of_square, then use this function in your implemented synced bn module.

**Comment by Spritea** [comment]
> @acgtyrant Is DistributedDataParallel working for you? Are you planning to send a Pull Request?
> 
> Thank you all for this beautiful work!

FYI, there is an open source repo created by NVIDIA, https://github.com/NVIDIA/apex, which supports SyncBN with `DistributedDataParallel` .

In fact, till now, it **only** supports SyncBN with `DistributedDataParallel`, and **doesn't support** SyncBN with `DataParallel`, see this issue. https://github.com/NVIDIA/apex/issues/115

**Comment by qchenclaire** [comment]
How about  using torch.nn.parallel.DistributedDataParallel but only running on one node with 8 gpus? Does this repo work?

**Comment by zhanghang1989** [comment]
> How about using torch.nn.parallel.DistributedDataParallel but only running on one node with 8 gpus? Does this repo work?

For `torch.nn.parallel.DistributedDataParallel` please use [torch.nn.SyncBatchNorm](https://pytorch.org/docs/master/nn.html?highlight=batchnorm#torch.nn.SyncBatchNorm)
